src/cargo/core/compiler/standard_lib.rs

```diff
@@ -158,6 +161,47 @@ pub fn resolve_std<'gctx>(
         HasDevUnits::No,
         crate::core::resolver::features::ForceAllTargets::No,
     )?;
+
+    // Verify that we have resolved to a subset of the lockfile
+    let lockfile = std::fs::read_to_string(&src_path.join("Cargo.lock"))
+        .expect("Couldn't read the Rust source's lockfile");
+
+    let encoded_lockfile: EncodableResolve = toml::from_str(&lockfile).unwrap();
+    let lockfile_packages = encoded_lockfile.package().expect("libstd has no packages!");
```

Comment:
The general practice I've heard and prefer for `expect` is it should be listing the invariant you are relying on